src/backends/perf_pt/decode.c

```diff
@@ -195,41 +196,63 @@ perf_pt_next_block(struct pt_block_decoder *decoder, int *decoder_status,
         panic("Unexpected decoder status: %d", *decoder_status);
     }
 
-    // Now fetch the block information.
+    // The libipt block decoder may return a partial block (it could have been
+    // interrupted for example). We abstract this detail away. Using a loop we
+    // record (and eventually return) the address of the first block we see,
+    // then keep decoding more blocks until we see a properly terminated block.
     struct pt_block block;
-    *decoder_status = pt_blk_next(decoder, &block, sizeof(block));
-    // Other +ve decoder status codes can arise here. We ignore them for now,
-    // and let them be detected by handle_events() above when we are next
-    // called.
-    if (*decoder_status < 0) {
-        if (*decoder_status == -pte_eos) {
-            // End of stream is flagged as an error in the case of pt_blk_next().
+    block.iclass = ptic_other;
+    bool first_block = true;
+    while (!block_is_terminated(&block)) {
+        if (handle_events(decoder, decoder_status, err) != true) {
+            // handle_events will have already called perf_pt_set_err().
+            return false;
+        } else if (*decoder_status & pts_eos) {
+            // End of stream.
             *addr = 0;
             return true;
-        } else {
-            // A real error.
-            perf_pt_set_err(err, perf_pt_cerror_ipt, -*decoder_status);
-            return false;
         }
-    }
-    // It's possible at this point that we get notified of an event in the
-    // stream. This will be handled in the next call to `perf_pt_next_block`.
-    if ((*decoder_status != 0) && (*decoder_status != pts_event_pending)) {
-        panic("Unexpected decoder status: %d", *decoder_status);
-    }
+        // It's possible at this point that we get notified of an event in the
+        // stream. This will be handled in the next call to `perf_pt_next_block`.
+        if ((*decoder_status != 0) && (*decoder_status != pts_event_pending)) {
+            panic("Unexpected decoder status: %d", *decoder_status);
+        }
 
-    // XXX A truncated block occurs when a block straddles a section boundary.
-    // In this case we may need some extra logic, but this should be rare.
-    if (block.truncated != 0) {
-        panic("Truncated blocks are not implemented");
-    }
+        // Now fetch the block information.
```

Comment:
Redundant comment.

Reply:
It was there before. But I agree we can kill it.